sysvar: add system variable `tidb_opt_enable_hash_join` (#14905)
pingcap/docs: system-variables.md

- Default value: `ON`
- This variable is used to control whether the optimizer estimates the number of rows based on column order correlation

### tidb_opt_enable_hash_join

- Scope: SESSION | GLOBAL

> - Persists to cluster: Yes

Reviewer: ```suggestion
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
```

- Default value: `ON`
- This variable is used to control whether the optimizer estimates the number of rows based on column order correlation

### tidb_opt_enable_hash_join

- Scope: SESSION | GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Boolean
- Default value: `ON`

> - With tidb_opt_enable_hash_join=off, the optimizer ignores hash join when generating execution plans.

Reviewer 2: ```suggestion
- With `tidb_opt_enable_hash_join=off`, the optimizer ignores hash join when generating execution plans.
```

- Default value: `ON`
- This variable is used to control whether the optimizer estimates the number of rows based on column order correlation

### tidb_opt_enable_hash_join <span class="version-mark">New in v7.4.0</span>

- Scope: SESSION | GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Boolean
- Default value: `ON`

> - This variable is used to control whether the optimizer will select hash joins for tables. The value is `ON` by default. If it is set to `OFF`, the optimizer ignores hash join when generating execution plans.

Reviewer 3: ```suggestion
- This variable is used to control whether the optimizer selects hash joins for tables. The value is `ON` by default. If it is set to `OFF`, the optimizer avoids selecting hash joins when generating execution plans, unless no other join algorithm is available.
```

> - If both the system variable `tidb_opt_enable_hash_join` and the `HASH_JOIN` hint are configured, the `HASH_JOIN` hint takes precedence. Even if `tidb_opt_enable_hash_join` is set to `off`, when you specify a `HASH_JOIN` hint in a query, the TiDB optimizer still enforces a hash join plan.

Reviewer 3: ```suggestion
- If both the system variable `tidb_opt_enable_hash_join` and the `HASH_JOIN` hint are configured, the `HASH_JOIN` hint takes precedence. Even if `tidb_opt_enable_hash_join` is set to `OFF`, when you specify a `HASH_JOIN` hint in a query, the TiDB optimizer still enforces a hash join plan.
```
